Cleanup stale gateway routes and tunnel ports on Agent startup

antoninbas opened the pull request:
In case of Agent restart, Nodes may have left the cluster while the
Agent was offline, and some gateway routes / tunnel ports may be
stale.

Even though stale routes might not create connectivity issue, they may
confuse users trying to troubleshoot networking issues, so it's better
to remove them.

Stale tunnel ports (either because a Node has left the cluster of
because the configuration is no longer valid) must be removed / updated
as well.

Another approach may be to use Resource Version when starting the watch
in the controller, as "delete" events would then be replayed, but that
would require us to store some state (latest Resource Version processed
by the controller) under /var/run and may make the main controller logic
more complex. Still I would be open to that approach as well.

Fixes #36

tnqn commented on test/e2e/basic_test.go:
```diff
@@ -16,11 +16,14 @@ package e2e
 	"time"
 
 	"github.com/vmware-tanzu/antrea/pkg/agent/util"
+	"k8s.io/apimachinery/pkg/util/wait"
```
put it to a separate group?

tnqn commented on test/e2e/framework.go:
```diff
@@ -50,9 +50,12 @@ const nameSuffixLength int = 8
+// running.
 const AntreaNamespace string = "kube-system"
 
+const AntreaGwName string = "gw0"
```
AntreaGWName?

tnqn commented on test/e2e/basic_test.go:
```diff
@@ -221,3 +224,158 @@ func TestIPAMRestart(t *testing.T) {
+
+	type Route struct {
+		peerPodCIDR *net.IPNet
+		peerPodGw   net.IP
```
peerPodGW?

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -126,35 +129,104 @@ func (c *Controller) enqueueNode(obj interface{}) {
+// removeStaleGatewayRoutes removes all the gateway routes which no longer correspond to a Node in
+// the cluster. If the antrea agent restarts and Nodes have left the cluster, this function will
+// take care of removing routes which are no longer valid.
+func (c *Controller) removeStaleGatewayRoutes() error {
```
I merged the IPSec PR. Seems you should remove IPSec tunnel ports as well. Maybe call deleteNodeRoute for the deleted Nodes.

antoninbas replied:
will update the PR

antoninbas commented:
@jianjuns could you take another look at this? I added support for removing stale tunnel ports.

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
+// changed, the tunnel port will also be deleted (the controller loop will later take care of
+// re-creating the port with the correct configuration).
+func (c *Controller) removeStaleTunnelPorts() error {
+	nodes, err := c.nodeLister.List(labels.Everything())
```
Does this always call to K8s API? If yes, probably makes sense to share a single call with removeStaleGatewayRoutes.

antoninbas replied:
I'm pretty sure it doesn't. It uses the local client cache.

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
+				continue
+			}
+
+			portName := util.GenerateNodeTunnelInterfaceName(node.Name)
```
I think you should use GenerateNodeTunnelInterfaceKey() here, as InterfaceName no more matches the interface store map key now.

antoninbas replied:
yes, you're correct, that's a typo

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
+
+	// remove all ports which are no longer needed or for which the configuration is no longer
+	// valid.
+	for _, ifaceID := range knownInterfaces {
```
To be clear, I suggest to use ifaceKey.

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
+			klog.Errorf("Interface %s can no longer be found in the interface store", ifaceID)
+			continue
+		}
+		if interfaceConfig.OFPort == types.DefaultTunOFPort {
```
OFPort should also work, but I would prefer to compare interface key or name here.

Do you want to cover removing the default tunnel port when IPSec is enabled?

antoninbas replied:
I thought about it which is why I brought it up yesterday when we chatted. However, I think it should be done in agent.go, which is where the code in charge of creating the default tunnel port is located. I opened a "low-priority" issue to track this: https://github.com/vmware-tanzu/antrea/issues/300

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -144,35 +147,178 @@ func (c *Controller) enqueueNode(obj interface{}) {
+		if err := c.ovsBridgeClient.DeletePort(interfaceConfig.PortUUID); err != nil {
+			klog.Errorf("Failed to delete OVS tunnel port %s: %v", interfaceConfig.InterfaceName, err)
+		}
+		c.interfaceStore.DeleteInterface(interfaceConfig)
```
If deleting OVS port failed, better to keep the interface in cache?

jianjuns replied:
Humm.. then we might still need to handle the tunnel type/Node change in createIPSecTunnelPort(). Maybe keep the TODO there?

antoninbas commented:
/test-all

antoninbas commented:
@jianjuns I have addressed your comments and added a new e2e test to check that tunnel ports can be removed properly.

While testing I noticed that deleting an Antrea agent Pod could take a long time because the `start_ovs_ipsec` script was ignoring the signal for up to 30 seconds, so I made the same changes to the script as the ones I made to `start_ovs` a while back. It cut testing time for the IPsec tests.

jianjuns commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -144,35 +147,179 @@ func (c *Controller) enqueueNode(obj interface{}) {
+				continue
+			}
+
+			portName := util.GenerateNodeTunnelInterfaceKey(node.Name)
```
Probably rename it to interfaceKey or ID.

antoninbas replied:
Done.

jianjuns commented on test/e2e/ipsec_test.go:
```diff
@@ -43,3 +48,65 @@ func TestIPSecTunnelConnectivity(t *testing.T) {
+	defer teardownTest(t, data)
+
+	t.Logf("Redeploy Antrea with IPSec tunnel enabled")
+	data.redeployAntrea(t, true)
```
How long it takes to redeploy Antrea? I wonder if the e2e tests will last too long if we continue to add such tests.

antoninbas replied:
After the fix to the `start_ovs_ipsec` it went from 80s to 30s. I think as long as the tests take about the same amount of time as the conformance / NP tests, we are good. We still have a few minutes of margin I believe.

antoninbas commented:
/test-all

antoninbas commented:
/test-e2e

antoninbas commented:
/test-e2e

antoninbas commented:
/test-e2e

antoninbas commented:
/test-e2e

antoninbas commented:
/test-e2e

antoninbas commented:
/test-all

antoninbas commented:
/test-all

antoninbas commented:
@jianjuns could I get a new approval on this? I had to rebase and fix a failing test.
